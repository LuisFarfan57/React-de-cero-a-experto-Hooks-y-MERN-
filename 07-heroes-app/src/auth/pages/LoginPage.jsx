import React, { useContext } from 'react'
import { useNavigate } from 'react-router-dom'
import { AuthContext } from '../context'

export const LoginPage = () => {
  const { login } = useContext(AuthContext)
  const navigate = useNavigate()

  const handleLogin = () => {
    login('Luis Farfan')

    // Checks if the lastPath is in the local storage, lo redirect to that page
    const lastPath = localStorage.getItem('lastPath') || '/marvel'

    navigate(lastPath, {
      replace: true
    })
  }

  return (
    <div className="container mt-5">
      <h1>Login</h1>
      <hr />

      <button className='btn btn-primary' onClick={handleLogin}>Login</button>
    </div>
  )
}
