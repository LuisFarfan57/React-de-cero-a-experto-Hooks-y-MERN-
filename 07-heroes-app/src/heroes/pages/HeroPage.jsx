import React, { useMemo } from 'react'
import { Navigate, useNavigate, useParams } from 'react-router-dom'
import { getHeroById } from '../helpers/getHeroById'
import { getHeroImgUrl } from '../helpers/getHeroImgUrl'

export const HeroPage = () => {
  const navigate = useNavigate()
  const { id } = useParams()
  const hero = useMemo(() => getHeroById(id), [id])

  if (!hero) return <Navigate to={'/marvel'} />

  const {
    superhero,
    publisher,
    alter_ego,
    first_appearance,
    characters
  } = hero

  const heroImgUrl = useMemo(() => getHeroImgUrl(id), [id])

  const handleNavigateBack = () => {
    navigate(publisher === 'DC Comics' ? '/dc' : '/marvel')
  }

  return (
    <div className="row mt-5">
      <div className="col-4">
        <img src={heroImgUrl} alt={superhero} className="img-thumbnail animate__animated animate__fadeInLeft" />
      </div>

      <div className='col-8'>
        <h3>{superhero}</h3>
        <ul className='list-group list-group-flush'>
          <li className='list-group-item'><b>Alter ego:</b> {alter_ego}</li>
          <li className='list-group-item'><b>Publisher:</b> {publisher}</li>
          <li className='list-group-item'><b>First appearance:</b> {first_appearance}</li>
        </ul>

        <h5 className='mt-3'>Characters</h5>
        <p>{characters}</p>

        <button className='btn btn-outline-primary' onClick={handleNavigateBack}>Go back</button>
      </div>
    </div>
  )
}
