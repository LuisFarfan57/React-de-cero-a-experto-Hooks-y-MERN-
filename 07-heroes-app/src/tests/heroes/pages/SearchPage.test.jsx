import { fireEvent, render, screen } from "@testing-library/react"
import { MemoryRouter, useNavigate } from "react-router-dom"
import { SearchPage } from "../../../heroes/pages/SearchPage"

const mockedUseNavigate = jest.fn()

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useNavigate: () => mockedUseNavigate
}))

describe('tests in <SearchPage /> component', () => {
  beforeEach(() => jest.clearAllMocks())

  it('shows correctly with the default values', () => {
    const { container } = render(
      <MemoryRouter>
        <SearchPage />
      </MemoryRouter>
    )

    expect(container).toMatchSnapshot()
  })

  it('should show Batman and the input with the batman value', () => {
    render(
      <MemoryRouter initialEntries={['/search?q=batman']}>
        <SearchPage />
      </MemoryRouter>
    )

    const input = screen.getByRole('textbox')

    expect(input.value).toBe('batman')
    expect(screen.getByRole('img').src).toContain('/assets/heroes/dc-batman.jpg')
    expect(screen.getByLabelText('searchHeroDiv').style.display).toBe('none')
  })

  it('should show an error if the hero is not found', () => {
    render(
      <MemoryRouter initialEntries={['/search?q=batman12345']}>
        <SearchPage />
      </MemoryRouter>
    )

    expect(screen.getByLabelText('noHeroDiv').style.display).toBe('')
  })

  it('should call the navigate when the user enter a hero name with the hero name', () => {
    render(
      <MemoryRouter initialEntries={['/search']}>
        <SearchPage />
      </MemoryRouter>
    )

    const input = screen.getByRole('textbox')
    fireEvent.change(input, { target: { value: 'Superman' } })
    const form = screen.getByLabelText('form')
    fireEvent.submit(form)

    expect(mockedUseNavigate).toHaveBeenCalledWith('?q=Superman')
  })
})
