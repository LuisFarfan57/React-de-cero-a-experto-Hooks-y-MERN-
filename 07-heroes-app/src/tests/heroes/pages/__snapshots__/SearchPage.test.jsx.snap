// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests in <SearchPage /> component shows correctly with the default values 1`] = `
<div>
  <h1>
    Search
  </h1>
  <hr />
  <div
    class="row"
  >
    <div
      class="col-5"
    >
      <h4>
        Searching
      </h4>
      <hr />
      <form
        aria-label="form"
      >
        <input
          autocomplete="off"
          class="form-control"
          name="searchText"
          placeholder="Search a hero"
          type="text"
          value=""
        />
        <button
          class="btn btn-outline-primary mt-3"
        >
          Search
        </button>
      </form>
    </div>
    <div
      class="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        aria-label="searchHeroDiv"
        class="alert alert-primary animate__animated animate__fadeIn"
      >
        Search a hero
      </div>
      <div
        aria-label="noHeroDiv"
        class="alert alert-danger animate__animated animate__fadeIn"
        style="display: none;"
      >
        No hero with 
        <b />
      </div>
      <div
        class="card animate__animated animate__fadeIn"
      >
        <div
          class="row no-gutters"
        >
          <div
            class="col-4"
          >
            <img
              alt="Batman"
              class="card-img"
              src="/assets/heroes/dc-batman.jpg"
            />
          </div>
          <div
            class="col-8"
          >
            <div
              class="card-body"
            >
              <h5
                class="card-title"
              >
                Batman
              </h5>
              <p
                class="card-text"
              >
                Bruce Wayne
              </p>
              <p
                class="card-text"
              >
                <small
                  class="text-muted"
                >
                  Detective Comics #27
                </small>
              </p>
              <a
                href="/hero/dc-batman"
              >
                More information
              </a>
            </div>
          </div>
        </div>
      </div>
      <div
        class="card animate__animated animate__fadeIn"
      >
        <div
          class="row no-gutters"
        >
          <div
            class="col-4"
          >
            <img
              alt="Superman"
              class="card-img"
              src="/assets/heroes/dc-superman.jpg"
            />
          </div>
          <div
            class="col-8"
          >
            <div
              class="card-body"
            >
              <h5
                class="card-title"
              >
                Superman
              </h5>
              <p
                class="card-text"
              >
                Kal-El
              </p>
              <p
                class="card-text"
              >
                <small
                  class="text-muted"
                >
                  Action Comics #1
                </small>
              </p>
              <a
                href="/hero/dc-superman"
              >
                More information
              </a>
            </div>
          </div>
        </div>
      </div>
      <div
        class="card animate__animated animate__fadeIn"
      >
        <div
          class="row no-gutters"
        >
          <div
            class="col-4"
          >
            <img
              alt="Flash"
              class="card-img"
              src="/assets/heroes/dc-flash.jpg"
            />
          </div>
          <div
            class="col-8"
          >
            <div
              class="card-body"
            >
              <h5
                class="card-title"
              >
                Flash
              </h5>
              <p
                class="card-text"
              >
                Jay Garrick
              </p>
              <p
                class="card-text"
              >
                Jay Garrick, Barry Allen, Wally West, Bart Allen
              </p>
              <p
                class="card-text"
              >
                <small
                  class="text-muted"
                >
                  Flash Comics #1
                </small>
              </p>
              <a
                href="/hero/dc-flash"
              >
                More information
              </a>
            </div>
          </div>
        </div>
      </div>
      <div
        class="card animate__animated animate__fadeIn"
      >
        <div
          class="row no-gutters"
        >
          <div
            class="col-4"
          >
            <img
              alt="Green Lantern"
              class="card-img"
              src="/assets/heroes/dc-green.jpg"
            />
          </div>
          <div
            class="col-8"
          >
            <div
              class="card-body"
            >
              <h5
                class="card-title"
              >
                Green Lantern
              </h5>
              <p
                class="card-text"
              >
                Alan Scott
              </p>
              <p
                class="card-text"
              >
                Alan Scott, Hal Jordan, Guy Gardner, John Stewart, Kyle Raynor, Jade, Sinestro, Simon Baz
              </p>
              <p
                class="card-text"
              >
                <small
                  class="text-muted"
                >
                  All-American Comics #16
                </small>
              </p>
              <a
                href="/hero/dc-green"
              >
                More information
              </a>
            </div>
          </div>
        </div>
      </div>
      <div
        class="card animate__animated animate__fadeIn"
      >
        <div
          class="row no-gutters"
        >
          <div
            class="col-4"
          >
            <img
              alt="Green Arrow"
              class="card-img"
              src="/assets/heroes/dc-arrow.jpg"
            />
          </div>
          <div
            class="col-8"
          >
            <div
              class="card-body"
            >
              <h5
                class="card-title"
              >
                Green Arrow
              </h5>
              <p
                class="card-text"
              >
                Oliver Queen
              </p>
              <p
                class="card-text"
              >
                <small
                  class="text-muted"
                >
                  More Fun Comics #73
                </small>
              </p>
              <a
                href="/hero/dc-arrow"
              >
                More information
              </a>
            </div>
          </div>
        </div>
      </div>
      <div
        class="card animate__animated animate__fadeIn"
      >
        <div
          class="row no-gutters"
        >
          <div
            class="col-4"
          >
            <img
              alt="Wonder Woman"
              class="card-img"
              src="/assets/heroes/dc-wonder.jpg"
            />
          </div>
          <div
            class="col-8"
          >
            <div
              class="card-body"
            >
              <h5
                class="card-title"
              >
                Wonder Woman
              </h5>
              <p
                class="card-text"
              >
                Princess Diana
              </p>
              <p
                class="card-text"
              >
                <small
                  class="text-muted"
                >
                  All Star Comics #8
                </small>
              </p>
              <a
                href="/hero/dc-wonder"
              >
                More information
              </a>
            </div>
          </div>
        </div>
      </div>
      <div
        class="card animate__animated animate__fadeIn"
      >
        <div
          class="row no-gutters"
        >
          <div
            class="col-4"
          >
            <img
              alt="Martian Manhunter"
              class="card-img"
              src="/assets/heroes/dc-martian.jpg"
            />
          </div>
          <div
            class="col-8"
          >
            <div
              class="card-body"
            >
              <h5
                class="card-title"
              >
                Martian Manhunter
              </h5>
              <p
                class="card-text"
              >
                J'onn J'onzz
              </p>
              <p
                class="card-text"
              >
                Martian Manhunter
              </p>
              <p
                class="card-text"
              >
                <small
                  class="text-muted"
                >
                  Detective Comics #225
                </small>
              </p>
              <a
                href="/hero/dc-martian"
              >
                More information
              </a>
            </div>
          </div>
        </div>
      </div>
      <div
        class="card animate__animated animate__fadeIn"
      >
        <div
          class="row no-gutters"
        >
          <div
            class="col-4"
          >
            <img
              alt="Robin/Nightwing"
              class="card-img"
              src="/assets/heroes/dc-robin.jpg"
            />
          </div>
          <div
            class="col-8"
          >
            <div
              class="card-body"
            >
              <h5
                class="card-title"
              >
                Robin/Nightwing
              </h5>
              <p
                class="card-text"
              >
                Dick Grayson
              </p>
              <p
                class="card-text"
              >
                <small
                  class="text-muted"
                >
                  Detective Comics #38
                </small>
              </p>
              <a
                href="/hero/dc-robin"
              >
                More information
              </a>
            </div>
          </div>
        </div>
      </div>
      <div
        class="card animate__animated animate__fadeIn"
      >
        <div
          class="row no-gutters"
        >
          <div
            class="col-4"
          >
            <img
              alt="Blue Beetle"
              class="card-img"
              src="/assets/heroes/dc-blue.jpg"
            />
          </div>
          <div
            class="col-8"
          >
            <div
              class="card-body"
            >
              <h5
                class="card-title"
              >
                Blue Beetle
              </h5>
              <p
                class="card-text"
              >
                Dan Garret
              </p>
              <p
                class="card-text"
              >
                Dan Garret, Ted Kord, Jaime Reyes
              </p>
              <p
                class="card-text"
              >
                <small
                  class="text-muted"
                >
                  Mystery Men Comics #1
                </small>
              </p>
              <a
                href="/hero/dc-blue"
              >
                More information
              </a>
            </div>
          </div>
        </div>
      </div>
      <div
        class="card animate__animated animate__fadeIn"
      >
        <div
          class="row no-gutters"
        >
          <div
            class="col-4"
          >
            <img
              alt="Black Canary"
              class="card-img"
              src="/assets/heroes/dc-black.jpg"
            />
          </div>
          <div
            class="col-8"
          >
            <div
              class="card-body"
            >
              <h5
                class="card-title"
              >
                Black Canary
              </h5>
              <p
                class="card-text"
              >
                Dinah Drake
              </p>
              <p
                class="card-text"
              >
                Dinah Drake, Dinah Lance
              </p>
              <p
                class="card-text"
              >
                <small
                  class="text-muted"
                >
                  Flash Comics #86
                </small>
              </p>
              <a
                href="/hero/dc-black"
              >
                More information
              </a>
            </div>
          </div>
        </div>
      </div>
      <div
        class="card animate__animated animate__fadeIn"
      >
        <div
          class="row no-gutters"
        >
          <div
            class="col-4"
          >
            <img
              alt="Spider Man"
              class="card-img"
              src="/assets/heroes/marvel-spider.jpg"
            />
          </div>
          <div
            class="col-8"
          >
            <div
              class="card-body"
            >
              <h5
                class="card-title"
              >
                Spider Man
              </h5>
              <p
                class="card-text"
              >
                Peter Parker
              </p>
              <p
                class="card-text"
              >
                <small
                  class="text-muted"
                >
                  Amazing Fantasy #15
                </small>
              </p>
              <a
                href="/hero/marvel-spider"
              >
                More information
              </a>
            </div>
          </div>
        </div>
      </div>
      <div
        class="card animate__animated animate__fadeIn"
      >
        <div
          class="row no-gutters"
        >
          <div
            class="col-4"
          >
            <img
              alt="Captain America"
              class="card-img"
              src="/assets/heroes/marvel-captain.jpg"
            />
          </div>
          <div
            class="col-8"
          >
            <div
              class="card-body"
            >
              <h5
                class="card-title"
              >
                Captain America
              </h5>
              <p
                class="card-text"
              >
                Steve Rogers
              </p>
              <p
                class="card-text"
              >
                <small
                  class="text-muted"
                >
                  Captain America Comics #1
                </small>
              </p>
              <a
                href="/hero/marvel-captain"
              >
                More information
              </a>
            </div>
          </div>
        </div>
      </div>
      <div
        class="card animate__animated animate__fadeIn"
      >
        <div
          class="row no-gutters"
        >
          <div
            class="col-4"
          >
            <img
              alt="Iron Man"
              class="card-img"
              src="/assets/heroes/marvel-iron.jpg"
            />
          </div>
          <div
            class="col-8"
          >
            <div
              class="card-body"
            >
              <h5
                class="card-title"
              >
                Iron Man
              </h5>
              <p
                class="card-text"
              >
                Tony Stark
              </p>
              <p
                class="card-text"
              >
                <small
                  class="text-muted"
                >
                  Tales of Suspense #39
                </small>
              </p>
              <a
                href="/hero/marvel-iron"
              >
                More information
              </a>
            </div>
          </div>
        </div>
      </div>
      <div
        class="card animate__animated animate__fadeIn"
      >
        <div
          class="row no-gutters"
        >
          <div
            class="col-4"
          >
            <img
              alt="Thor"
              class="card-img"
              src="/assets/heroes/marvel-thor.jpg"
            />
          </div>
          <div
            class="col-8"
          >
            <div
              class="card-body"
            >
              <h5
                class="card-title"
              >
                Thor
              </h5>
              <p
                class="card-text"
              >
                Thor Odinson
              </p>
              <p
                class="card-text"
              >
                <small
                  class="text-muted"
                >
                  Journey into Myster #83
                </small>
              </p>
              <a
                href="/hero/marvel-thor"
              >
                More information
              </a>
            </div>
          </div>
        </div>
      </div>
      <div
        class="card animate__animated animate__fadeIn"
      >
        <div
          class="row no-gutters"
        >
          <div
            class="col-4"
          >
            <img
              alt="Hulk"
              class="card-img"
              src="/assets/heroes/marvel-hulk.jpg"
            />
          </div>
          <div
            class="col-8"
          >
            <div
              class="card-body"
            >
              <h5
                class="card-title"
              >
                Hulk
              </h5>
              <p
                class="card-text"
              >
                Bruce Banner
              </p>
              <p
                class="card-text"
              >
                <small
                  class="text-muted"
                >
                  The Incredible Hulk #1
                </small>
              </p>
              <a
                href="/hero/marvel-hulk"
              >
                More information
              </a>
            </div>
          </div>
        </div>
      </div>
      <div
        class="card animate__animated animate__fadeIn"
      >
        <div
          class="row no-gutters"
        >
          <div
            class="col-4"
          >
            <img
              alt="Wolverine"
              class="card-img"
              src="/assets/heroes/marvel-wolverine.jpg"
            />
          </div>
          <div
            class="col-8"
          >
            <div
              class="card-body"
            >
              <h5
                class="card-title"
              >
                Wolverine
              </h5>
              <p
                class="card-text"
              >
                James Howlett
              </p>
              <p
                class="card-text"
              >
                <small
                  class="text-muted"
                >
                  The Incredible Hulk #180
                </small>
              </p>
              <a
                href="/hero/marvel-wolverine"
              >
                More information
              </a>
            </div>
          </div>
        </div>
      </div>
      <div
        class="card animate__animated animate__fadeIn"
      >
        <div
          class="row no-gutters"
        >
          <div
            class="col-4"
          >
            <img
              alt="Daredevil"
              class="card-img"
              src="/assets/heroes/marvel-daredevil.jpg"
            />
          </div>
          <div
            class="col-8"
          >
            <div
              class="card-body"
            >
              <h5
                class="card-title"
              >
                Daredevil
              </h5>
              <p
                class="card-text"
              >
                Matthew Michael Murdock
              </p>
              <p
                class="card-text"
              >
                <small
                  class="text-muted"
                >
                  Daredevil #1
                </small>
              </p>
              <a
                href="/hero/marvel-daredevil"
              >
                More information
              </a>
            </div>
          </div>
        </div>
      </div>
      <div
        class="card animate__animated animate__fadeIn"
      >
        <div
          class="row no-gutters"
        >
          <div
            class="col-4"
          >
            <img
              alt="Hawkeye"
              class="card-img"
              src="/assets/heroes/marvel-hawkeye.jpg"
            />
          </div>
          <div
            class="col-8"
          >
            <div
              class="card-body"
            >
              <h5
                class="card-title"
              >
                Hawkeye
              </h5>
              <p
                class="card-text"
              >
                Clinton Francis Barton
              </p>
              <p
                class="card-text"
              >
                <small
                  class="text-muted"
                >
                  Tales of Suspense #57
                </small>
              </p>
              <a
                href="/hero/marvel-hawkeye"
              >
                More information
              </a>
            </div>
          </div>
        </div>
      </div>
      <div
        class="card animate__animated animate__fadeIn"
      >
        <div
          class="row no-gutters"
        >
          <div
            class="col-4"
          >
            <img
              alt="Cyclops"
              class="card-img"
              src="/assets/heroes/marvel-cyclops.jpg"
            />
          </div>
          <div
            class="col-8"
          >
            <div
              class="card-body"
            >
              <h5
                class="card-title"
              >
                Cyclops
              </h5>
              <p
                class="card-text"
              >
                Scott Summers
              </p>
              <p
                class="card-text"
              >
                <small
                  class="text-muted"
                >
                  X-Men #1
                </small>
              </p>
              <a
                href="/hero/marvel-cyclops"
              >
                More information
              </a>
            </div>
          </div>
        </div>
      </div>
      <div
        class="card animate__animated animate__fadeIn"
      >
        <div
          class="row no-gutters"
        >
          <div
            class="col-4"
          >
            <img
              alt="Silver Surfer"
              class="card-img"
              src="/assets/heroes/marvel-silver.jpg"
            />
          </div>
          <div
            class="col-8"
          >
            <div
              class="card-body"
            >
              <h5
                class="card-title"
              >
                Silver Surfer
              </h5>
              <p
                class="card-text"
              >
                Norrin Radd
              </p>
              <p
                class="card-text"
              >
                <small
                  class="text-muted"
                >
                  The Fantastic Four #48
                </small>
              </p>
              <a
                href="/hero/marvel-silver"
              >
                More information
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
